---
### **Window function**

---

In the previouse sections, we have learnt about grouping data, grouping happens after after FROM/WHERE, we learnt about the having keyword that it is special and its a filter for groups. 

We learnt about grouping set and rollups that they are useful for multiple groupings in a single query 

We learnt that grouping data is not a silver bullet meaning we can't answer all questions with it

---

So what are we missing?

How do we apply functions against rows related to the current row?

How do we take the calculation that we can do against a group and show it against the current row?

Let's put it in an example

What if we wanted to know the average salary per department 

Let's do it

---

In [1]:
%load_ext sql

In [2]:
%sql postgresql://postgres:Kithusi254@localhost:5432/Employees

'Connected: postgres@Employees'

In [3]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [4]:
%%sql

--What is the average salary per department

SELECT  emp.emp_no,
        CONCAT(first_name, ' ', last_name)
        AS "name",
        dep.dept_name,
        dpt.dept_no,
        sa.salary,
        ROUND(AVG(sa.salary) OVER(
                PARTITION BY dep.dept_name
        ), 2) 
        AS "AVG salary per dep"
FROM salaries AS sa
INNER JOIN employees AS emp USING(emp_no)
INNER JOIN dept_emp AS dpt USING(emp_no)
INNER JOIN departments AS dep USING(dept_no)
--GROUP BY dep.dept_name, dpt.dept_no
ORDER BY emp.emp_no
LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
50 rows affected.


emp_no,name,dept_name,dept_no,salary,AVG salary per dep
10001,Georgi Facello,Development,d005,74333,59478.90
10001,Georgi Facello,Development,d005,75286,59478.90
10001,Georgi Facello,Development,d005,66074,59478.90
10001,Georgi Facello,Development,d005,62102,59478.90
10001,Georgi Facello,Development,d005,66596,59478.90
10001,Georgi Facello,Development,d005,81025,59478.90
10001,Georgi Facello,Development,d005,66961,59478.90
10001,Georgi Facello,Development,d005,85112,59478.90
10001,Georgi Facello,Development,d005,88958,59478.90
10001,Georgi Facello,Development,d005,75994,59478.90


---

That was easy, but what if we wanted to add the average to every salary so we could visually see how much each employee is from the everage?

We have calculated the average and we know the average per department, but let's say we wanted to do a query and in that query we wanted to see the salary of employee x and next to it the average for salaries in his or her department right beside there salary and we can see how far they are from the average

We can achieve this through the window function

---

#### **Window function**

Our above problem statement was what if i wanted to show something beside my row based on a group calculation

**Window functions create a new column based on functions performed on a subset or "window" of data**

So we are applying a function against a subset or window of the data - the data that we are requesting in our query

**A window function can only be applied against the window of data that we have requested** 

So the window of data we've requested is the window upon which we can apply a window function to perform a calculation to give us an answer.

This becomes extreamly important when you wanna do data analysis against some of the data that is returned 

Window function have many applications and if your job is doing data analysis, you may find yourself using window function alot in your day to day query

SYNTAX

> **WINDOW_FUNCTION(arg1, arg2,..) OVER ([PARTITION BY partition_expression] [ORDER BY sort_expression [ASC|DESC][NULLS {FIRST | LAST}]])**

The syntax looks something like the above.

So, how do we apply it in real life?

We have a function called window function and it can potentially take arguments, its a function that takes arguments 

**The key part here is that for a window function to be applied as a window function, we need the over keyword because its gonna tell us, heey, apply this function over this window os data this subset of data** 

This is very key to remember

And then optionaly, we can give it a partition, we can give an order, these are optional parameters and we are going to get into them later

If we just pass over, it will apply in the entire window of data we requested 

If we request 10 million records, the window function applies against each individual row

So each time that we apply a window function, its going to apply again and again, against that subset of data hence if you request 10 million records, we are going to apply the window function against those 10 million records 10 million times. Due to this, our query may take a very long time.

Window function can take longer to run because they perform against the whole window of data that we selected against each individual data in that window.

Let's say we wanted to apply this window function to get the max salary and then get it alongside each individual salary?

Let's do it  practicaly....

---

In [5]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [6]:
%%sql

--Max salary alongside each individual salary

SELECT  emp.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name)
        AS "name",
        sa.salary,
        MAX(salary) OVER()
        AS "Max salary"
FROM employees AS emp
INNER JOIN salaries AS sa USING(emp_no)
ORDER BY emp.emp_no
LIMIT(50);


 * postgresql://postgres:***@localhost:5432/Employees
50 rows affected.


emp_no,name,salary,Max salary
10001,Georgi Facello,60117,158220
10001,Georgi Facello,62102,158220
10001,Georgi Facello,66074,158220
10001,Georgi Facello,66596,158220
10001,Georgi Facello,66961,158220
10001,Georgi Facello,71046,158220
10001,Georgi Facello,74333,158220
10001,Georgi Facello,75286,158220
10001,Georgi Facello,75994,158220
10001,Georgi Facello,76884,158220


---

After running the above code, we can see that we are getting each individual salary and the highest salary being paid in the company, we are getting them side by side

Let's say we set a limit of 50 records, we want the first 50 records, the window function will apply against the whole salary - current window and then give us the first 50 records after apply the window function hence the max salary may also not be part of the first 50 records.

But when we apply a filter, the window function will work within that filter, for instance if we sa we want salaries below 70000, it will work against that windo of data hence the highest salary will come from that window of data, 70000 and below eg, you may get the highest salary as 69000 so long as it is below 70000.

So the window of data applies against filters but it does not aplly against limits

So if we say limit my data, cut it off at 100 it still gonna apply against the entire dataset however when we apply filters its applying against that windo of data

Its is very important to understand how queries execute in such instances because the window of data is directly related to whatever the query will return 

**The window of data refers to the query calculated out** 

When we are using limit, we are just saying, of all the data that we got, just give me this

Our queries can take longer to execute because as the window of data grows, as we return more rows to the window function it takes longer and longer to apply depending on the function we are trtying to apply against each and every individual row

Let's do it precticaly..

---

In [7]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;


 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [8]:
%%sql

--Max salary alongside each individual with a salary < 70000

SELECT  emp.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name),
        sa.salary,
        MAX(salary) OVER()
        AS "Max salary"
FROM employees AS emp
INNER JOIN salaries AS sa USING(emp_no)
WHERE sa.salary < 70000
ORDER BY emp.emp_no
LIMIT 50;


 * postgresql://postgres:***@localhost:5432/Employees
50 rows affected.


emp_no,concat,salary,Max salary
10001,Georgi Facello,60117,69999
10001,Georgi Facello,62102,69999
10001,Georgi Facello,66074,69999
10001,Georgi Facello,66596,69999
10001,Georgi Facello,66961,69999
10002,Bezalel Simmel,65828,69999
10002,Bezalel Simmel,65909,69999
10002,Bezalel Simmel,67534,69999
10002,Bezalel Simmel,69366,69999
10003,Parto Bamford,40006,69999


---

### **Partition By**

---

PARTITION BY is used to divide rows into groups to aplly the function against(optional)

Its optional

Let's do it practicaly...

---


In [9]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [10]:
%%sql

--AVG salary for each department alongside the salaries of each 
--individual

SELECT  emp.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name),
        emp.gender,
        dep.dept_no,
        dep.dept_name,
        sa.salary,
        ROUND(AVG(sa.salary) OVER(
            PARTITION BY dep.dept_name
        ), 2)
        AS "AVG salary per dep"

FROM employees AS emp
INNER JOIN salaries AS sa USING(emp_no)
INNER JOIN dept_emp AS dpt USING(emp_no)
INNER JOIN departments AS dep USING(dept_no)
ORDER By emp.emp_no
LIMIT 50;


 * postgresql://postgres:***@localhost:5432/Employees
50 rows affected.


emp_no,concat,gender,dept_no,dept_name,salary,AVG salary per dep
10001,Georgi Facello,M,d005,Development,76884,59478.90
10001,Georgi Facello,M,d005,Development,85097,59478.90
10001,Georgi Facello,M,d005,Development,60117,59478.90
10001,Georgi Facello,M,d005,Development,74333,59478.90
10001,Georgi Facello,M,d005,Development,81025,59478.90
10001,Georgi Facello,M,d005,Development,75994,59478.90
10001,Georgi Facello,M,d005,Development,85112,59478.90
10001,Georgi Facello,M,d005,Development,84917,59478.90
10001,Georgi Facello,M,d005,Development,75286,59478.90
10001,Georgi Facello,M,d005,Development,66596,59478.90


---

We can see how partitioning the data can help us draw more insights

Now we know, for this employee, based on the department there in, we will know the average salary for there department

By partitioning it in this way, we can now the conclusion of, heey, in this department you are this off of the average salary

This gives alot of insight into the way the business is operating

Are the salary compensation fare

We can see that our window function has generated the average salary per department and by partitioning, we have returned both the column that it was partitioned by and the average salary

Partitioning changes the lens of the window function to look at a partition, a subset of data, in our case by department name

It helps us to draw deeper conclusions, grouped conclusions

Through it, we can run more indepth analysis

---

In [11]:
%%sql

/*

Let's get the employees average salary based on departments alongside each each employee's salary
This time select the emp_no, salary, name, dpt no, average salary per department, and the 
department name.

*/

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema= 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [12]:
%%sql

SELECT 
        sa.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name) AS "name",
        dpt.dept_no,
        dep.dept_name,
        sa.salary,
        ROUND(AVG(sa.salary) OVER(
            PARTITION BY dept_name
        ), 2) AS "AVG salary per dep"

FROM salaries AS sa
INNER JOIN employees AS emp USING(emp_no)
INNER JOIN dept_emp AS dpt USING(emp_no)
INNER JOIN departments AS dep USING(dept_no) 
ORDER BY emp.emp_no
LIMIT 100;

 * postgresql://postgres:***@localhost:5432/Employees
100 rows affected.


emp_no,name,dept_no,dept_name,salary,AVG salary per dep
10001,Georgi Facello,d005,Development,80013,59478.90
10001,Georgi Facello,d005,Development,88958,59478.90
10001,Georgi Facello,d005,Development,85112,59478.90
10001,Georgi Facello,d005,Development,75994,59478.90
10001,Georgi Facello,d005,Development,76884,59478.90
10001,Georgi Facello,d005,Development,60117,59478.90
10001,Georgi Facello,d005,Development,85097,59478.90
10001,Georgi Facello,d005,Development,81025,59478.90
10001,Georgi Facello,d005,Development,81097,59478.90
10001,Georgi Facello,d005,Development,66074,59478.90


---

### **ORDER BY** 

---

ORDER BY is used to order the results within a window function

It operates strangly inside window function.

Let's visualize it to see what we mean

---


In [13]:
%%sql

SELECT  sa.emp_no,
        COUNT(sa.salary) OVER(
            ORDER BY emp_no
        )
FROM salaries AS sa
LIMIT 100;


 * postgresql://postgres:***@localhost:5432/Employees
100 rows affected.


emp_no,count
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17


---

In the Query above we are doing a cont of the salary as it will help us to visualize better what ever is happening.

After running the query above, we realize that it is 17 unlike the previouse query without the ORDER BY which would give us the total count yet it is not PARTITION BY clouse.

Such results we say when we were practicing PARTITION BY clouse which groups our data as we are observing now.

Why would it change the way that the window function occurs?

If we were to do a partition, should we be able to count what is in that partition?

Let's do it practically..

---

In [14]:
%%sql

SELECT  sa.emp_no,
        COUNT(sa.salary) OVER(
            PARTITION BY emp_no
        )
FROM salaries AS sa
LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
50 rows affected.


emp_no,count
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17


---

We observe the same thing because we are counting specifically how many salaries are there that emp no.

Let's go back to ORDER BY

---

In [15]:
%%sql

SELECT  sa.emp_no,
        COUNT(sa.salary) OVER(
            ORDER BY emp_no
        )
FROM salaries AS sa 
LIMIT 30;

 * postgresql://postgres:***@localhost:5432/Employees
30 rows affected.


emp_no,count
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17


---

But if you scroll down, you will realize that there is a difference. We are not counting as we would think, we are not doing it like PARTITION BY clouse would do and ORDER BY is not partition based.

You will oberve we are counting 17 salaries in group 1 and then 23 salaries in group 2

How is that possible?

It is clear that 10001 goes up to 17, so it makes sense they have 17 salaries but then 10002 goes from 18 to 23, hence has 6 salaries but we are getting 23.

After doing the math you will realize the order is cumulative, 17 plus 6 which gives us 23.

The question remains why would ORDER BY change the way that the window function is going to count?

**This is becoz ORDER BY has a special property called FRAMING**

It changes the FRAME of the window function.

When we use ORDER BY withing a WINDOW function, we are telling our query and the WINDOW function used withing that query take into acc everything before me and myself

So we are changing the lense at which the window function is looing at the data.

When we remove the ORDER BY, its clear that the window function is taking into account the entirety of the returned rows but when we add an ORDER BY clouse, it becomes clear that we are taking into account itself, so everything itself but then if we look at 10002, its taking account itself and everything that came before it, that is 10001

So it kinda works like cumulative sum

More explanation below...

---


---
Previousely, we've seen:

A window function says, For each row, look at a set of related rows and compute something

That set of related rows is called window

---

#### **How the window is built(3 layers)**

1. **PARTITION BY**

Splits data into groups, "Which rows belong together"

2. **ORDER BY**

Arranges the rows inside each group

" In what sequence should i process them?"

It introduces sequence

When we say that ORDER BY introduces sequence, we mean that it creates a processing order, like a line.

Example:

| salary |
| ------ |
| 3000   |
| 1000   |
| 2000   |

With:

* * **SUM (salary) OVER (ORDER BY salary)**

The sequence becomes:

1000 → 2000 → 3000

Another example (DESC)

* * **SUM(salary) OVER (ORDER BY salary DESC)**

The sequence becomes:

3000 → 2000 → 1000

That's it:

Sequence - the path postgres follows row by row

Not final display

Not grouping

Just the order of movement

If you get this, everything else (running totals, rank, etc.) becomes obvious 💯

**The BIG misunderstanding**

**Most people think ORDER BY controls the results but ORDER BY only defines the sequence**

---


---

### **FRAMING**

We have just gone through ORDER BY and we have seen that when we use it, it causes strange things to happen.

Those things happen because of what we call FRAMING in WINDOW function.

When we are trying to get deterministic results with our window function and we through ORDER BY into the mix, it may completely change the outcome of what we were originally doing and that has to do with something called FRAMING

**When using a FRAME clouse in a WINDOW function we can create a sub-range or FRAME**

**When we use PARTITION BY let's say emp_no, we are grouping the data logicaly by the emp_no but when we use ORDER BY, we are changing the FRAME at which we are looking at the data**

So, when we talk abou the FRAMING what we are talking about is the way that the window function interprets what it should do with the data.

Framing comes with a very specific syntax

![](frame.png)



And when we through all that together, we can say somthing like:

> **PARTITION BY category ORDER BY price RANGE BETWEEN UNBOUND PRECEEDING AND CURRENT FOW**

In the above line of code we are saying, create a PARTITION BY category, then ORDER it BY price, and that range of data that we specified in the start of the query, and look at everything that came before - UNBOUND PRECEDING, and the CURRENT ROW.

What we are trying to basically say, when we use ORDER BY, we are basically changing the FRAME of refference

When we don't use ORDER BY, by default the FRAME we look at is usually the entire partition and when we use OVER with out partition, the entire partition is the entire window.

When using OVER alone as we saw, we are runing calculations against all the data.

When we introduce ORDER BY, by default the FRAMING is usually everything before the current row and the current row.

Let's put it into practice...

---

In [16]:
%%sql

SELECT  sa.emp_no,
        COUNT(salary) OVER(
            ORDER BY emp_no
        )

FROM salaries AS sa 
LIMIT 30;

 * postgresql://postgres:***@localhost:5432/Employees
30 rows affected.


emp_no,count
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17


---

If we ORDER BY emp_no as we have done above, what we are going to see is a comulative sum where we are cumulatively counting everything that came before, with everything that comes after.

We get results we are getting because we ORDER BY emp_no. It is counting everything it comes across when it counts that specific emp_no, the amount of times it occurs basically and its adding that to the previouse sum.

But what if we apply it into a PARTITION BY?

It depends on the partition we are going to do.

Let's say we partition by emp_no

It is going to order within that specific groups for instance 10001, it will order 10001 cumulatively and once its done its done with 100001. It starts again in 10002, orders within that group and so on.

Its only going to look inside of that partition

Its going to act as if it were a normal partition by as we see below.

---

In [17]:
%%sql

SELECT  sa.emp_no,
        COUNT(sa.salary) OVER(
            PARTITION BY emp_no
            ORDER BY emp_no
        )
FROM salaries AS sa 
LIMIT 30;

 * postgresql://postgres:***@localhost:5432/Employees
30 rows affected.


emp_no,count
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17


---

Its going to add the first partition 10001 alone to get 17, and then count the second partition 10002 to get 6 as the count as we see above, its not going to add them cumulatively like it did without partition.

---

Let's say we ordered by salary....

---

In [18]:
%%sql

SELECT  sa.emp_no,
        COUNT(sa.salary) OVER(
            PARTITION BY sa.emp_no
            ORDER BY sa.salary
        )
FROM salaries AS sa 
LIMIT 30;

 * postgresql://postgres:***@localhost:5432/Employees
30 rows affected.


emp_no,count
10001,1
10001,2
10001,3
10001,4
10001,5
10001,6
10001,7
10001,8
10001,9
10001,10


---

Because we are partitioning by emp_no and ordering by salary, our partition becomes the emp_no but our order is the salary and because they are unique values, its going to look at it self and what came before it and its going to add it and becuase the values are unique, we are going to go from 1 upto 17 as we see above

If there were salaries which were the same, we would see the same value occur like we say when we ordered by emp_no
 
The way you partition and the way that you order, really influence each other 

When we partition by emp_no, we are taking into account a specific group but when we look at the order, it changes the way that we are going to look at that partition, **it changes the way that our function is going to interact with that data, our count is now cumulative, it taking into acc the previouse and the current**

**It is very important to know how FRAMING will change when we introduce an ORDER BY**

It completely changes the way the function will interact

Deterministic results that we got when we ran a partition, can completely change when we introduce an ORDER BY and that is because the way we look at the FRAM changes.

This type of syntax is used to change the way the interaction of the window function works, in our case COUNT.

ORDER BY works in a way that it is taking into acc the current fram and the previouse one but what if we wanted to change that

Let's do it practically...

---

In [19]:
%%sql

SELECT  sa.emp_no,
        COUNT(sa.salary) OVER(
            PARTITION BY emp_no
            ORDER BY salary
            RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
        )
FROM salaries AS sa 
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/Employees
20 rows affected.


emp_no,count
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17
10001,17


---

If we add:

> **RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING**

We see what we saw previously in the first query

There are alot of FRAMEs that we can play around with to see how our data will be affected

We can be able to run countless types of functions, AVGs, SUMs, we can plug in the last value or the first value, we can get the positioning. 

FRAMING becomes important when we are doing data analysis

FRAMING is important and we can chnage the framing a partition by using something like the above syntax

We can go to an extent of changing the frame without having to use order by and it works as if we were using an order by 

Let's do it practically....

---

In [20]:
%%sql

SELECT  sa.emp_no,
        COUNT(sa.salary) OVER(
            PARTITION BY emp_no
            ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
        )

FROM salaries AS sa
LIMIT 30;

 * postgresql://postgres:***@localhost:5432/Employees
30 rows affected.


emp_no,count
10001,1
10001,2
10001,3
10001,4
10001,5
10001,6
10001,7
10001,8
10001,9
10001,10


---

We see its operating as if it were using an ORDER BY.

The way that we interact with the FRAMING, is important, are we going to take into account the rows or the range are we going to look at the order by or are we going to change the way we look at the frame.

---

There are so many Window functions that we can use and some of them include:-

![](Screenshot_2026-04-01_15-02-14.png)




They include as we see in the screenshot above:

* **SUM/MIN/MAX/AVG**
* **FIRST_VALUE**
* **LAST_VALUE**
* **NTH_VALUE**
* **PERCENT_RANK**
* **RANK**
* **ROW_NUMBER**
* **LAG/LEG**

Majority of these functions are to be used when we are doing somekind of data analysis, predictive things, when we wanna ran somekind of calculation etc..

---

---

### **FIRST_VALUE**

It returns a value evaluated against the first row within its partition.

So if we have a partition or even the entire window, we are going to grab the first value.

Let's solve a problem to understand how LAST_VALUE works.

Let's say we wanted to get the first salary of each employee?

---

In [21]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name; 

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [22]:
%%sql

SELECT  DISTINCT emp.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name)
        AS "name",
        emp.gender,
        dpt.dept_no,
        dep.dept_name,
        FIRST_VALUE(sa.salary) OVER(
            PARTITION BY emp.emp_no
        )
        AS "current salary",
        FIRST_VALUE(sa.from_date) OVER(
            PARTITION BY emp.emp_no
        )
        AS "from_date",
        FIRST_VALUE(sa.to_date) OVER(
            PARTITION BY emp.emp_no
        )
        AS "to_date"
FROM employees AS emp
INNER JOIN salaries AS sa USING(emp_no)
INNER JOIN dept_emp AS dpt USING(emp_no)
INNER JOIN departments AS dep USING(dept_no)
LIMIT 50;



 * postgresql://postgres:***@localhost:5432/Employees
50 rows affected.


emp_no,name,gender,dept_no,dept_name,current salary,from_date,to_date
10001,Georgi Facello,M,d005,Development,60117,1986-06-26,1987-06-26
10002,Bezalel Simmel,F,d007,Sales,65828,1996-08-03,1997-08-03
10003,Parto Bamford,M,d004,Production,40006,1995-12-03,1996-12-02
10004,Chirstian Koblick,M,d004,Production,40054,1986-12-01,1987-12-01
10005,Kyoichi Maliniak,M,d003,Human Resources,78228,1989-09-12,1990-09-12
10006,Anneke Preusig,F,d005,Development,40000,1990-08-05,1991-08-05
10007,Tzvetan Zielinski,F,d008,Research,56724,1989-02-10,1990-02-10
10008,Saniya Kalloufi,M,d005,Development,46671,1998-03-11,1999-03-11
10009,Sumant Peac,F,d006,Quality Management,60929,1985-02-18,1986-02-18
10010,Duangkaew Piveteau,F,d004,Production,72488,1996-11-24,1997-11-24


---

### **LAST_VALUE**

Returns a value evaluated against the last row within its partition.

Let's do it practically to see how it works.

Let's get the current salary of each employee

---

In [23]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [24]:
%%sql

SELECT  DISTINCT emp.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name)
        AS "name",
        emp.gender,
        dpt.dept_no,
        dep.dept_name,
        LAST_VALUE(sa.salary) OVER(
            PARTITION BY emp.emp_no
        )
        AS "carrent salary",
        LAST_VALUE(sa.from_date) OVER(
            PARTITION BY emp.emp_no
        )
        AS "from_date",
        LAST_VALUE(sa.to_date) OVER(
            PARTITION BY emp.emp_no
        )
        AS "to_date"
FROM employees AS emp
INNER JOIN salaries AS sa USING(emp_no)
INNER JOIN dept_emp AS dpt USING(emp_no)
INNER JOIN departments AS dep USING(dept_no)
LIMIT 50;


 * postgresql://postgres:***@localhost:5432/Employees
50 rows affected.


emp_no,name,gender,dept_no,dept_name,carrent salary,from_date,to_date
10001,Georgi Facello,M,d005,Development,88958,2002-06-22,9999-01-01
10002,Bezalel Simmel,F,d007,Sales,72527,2001-08-02,9999-01-01
10003,Parto Bamford,M,d004,Production,43311,2001-12-01,9999-01-01
10004,Chirstian Koblick,M,d004,Production,74057,2001-11-27,9999-01-01
10005,Kyoichi Maliniak,M,d003,Human Resources,94692,2001-09-09,9999-01-01
10006,Anneke Preusig,F,d005,Development,59755,2001-08-02,9999-01-01
10007,Tzvetan Zielinski,F,d008,Research,88070,2002-02-07,9999-01-01
10008,Saniya Kalloufi,M,d005,Development,52668,2000-03-10,2000-07-31
10009,Sumant Peac,F,d006,Quality Management,94409,2002-02-14,9999-01-01
10010,Duangkaew Piveteau,F,d004,Production,80324,2001-11-23,9999-01-01


---

There are many ways to solve this problem some of them are easier and not just this problems, all the problems

There is no 100% correct way but there are better ways and there are worse ways but as long us you get the answer that you want, its all about improving as you learn

So what ever way you decide to solve something as long as we are getting the right answer, we'll build up towards that mastery level, understanding, and comparison of performance to heey, this is right, this is better, there is no wrong method hence there is only improvements to be made from there.

---

---

### **SUM - WINDOW FUNCTION**

Sums the values within a group depending on the framing.

It answers the quesion:

What if we wanted to get the sum of the values within  a group depending on the framing 

We wanna some something but we wanna do it depending on the partition and framing.

Let's do it practically to understand it

What if we wanted to see how much cumulatively a customer has ordered at our store?

So over time, how much the total becomes

Let's dive into the quering.....


In [25]:
%sql postgresql://postgres:Kithusi254@localhost:5432/Store

'Connected: postgres@Store'

In [26]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
8 rows affected.


table_name,string_agg
categories,"category , categoryname"
cust_hist,"customerid , orderid , prod_id"
customers,"customerid , firstname , lastname , address1 , address2 , city , state , zip , country , region , email , phone , creditcardtype , creditcard , creditcardexpiration , username , password , age , income , gender"
inventory,"prod_id , quan_in_stock , sales"
orderlines,"orderlineid , orderid , prod_id , quantity , orderdate"
orders,"orderid , orderdate , customerid , netamount , tax , totalamount"
products,"prod_id , category , title , actor , price , special , common_prod_id"
reorder,"prod_id , date_low , quan_low , date_reordered , quan_reordered , date_expected"


In [27]:
%%sql

--How cumulatively has a customer order from our store?

SELECT  ord.customerid,
        CONCAT(cst.firstname, ' ', cst.lastname)
        AS "name",
        cst.gender,
        cst.age,
        cst.email,
        cst.phone,
        cst.city,
        cst.country,
        cst.region,
        ord.orderid,
        ord.netamount,
        SUM(netamount) OVER(
            PARTITION BY ord.customerid
            ORDER BY ord.orderid
        ) AS "Cum sum"
FROM orders AS ord
INNER JOIN customers AS cst USING(customerid)
ORDER BY ord.customerid 
LIMIT 30;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
30 rows affected.


customerid,name,gender,age,email,phone,city,country,region,orderid,netamount,Cum sum
2,HQNMZH UNUKXHJVXB,M,80,UNUKXHJVXB@dell.com,5119315633,YNCERXJ,US,1,10677,5.08,5.08
3,JTNRNB LYYSHTQJRE,M,47,LYYSHTQJRE@dell.com,6297761196,LWVIFXJ,US,1,2337,39.06,39.06
6,FXDZBW BAXPEEKXVJ,M,72,BAXPEEKXVJ@dell.com,6192740010,OPLRCNT,US,1,9077,323.30,323.30
7,WVZTXZ RMEVXCQGQF,M,52,RMEVXCQGQF@dell.com,9743191382,SIIGBQF,US,1,6239,341.44,341.44
11,XQVVMI KRPGDBCQJH,M,58,KRPGDBCQJH@dell.com,2415449050,ICLYPGR,US,1,1187,285.39,285.39
12,KGISQZ IXDKAUUHCW,F,27,IXDKAUUHCW@dell.com,1896033667,SBLZSFM,US,1,3710,350.87,350.87
13,LURLDP PNPJHXMEPN,M,43,PNPJHXMEPN@dell.com,3029418206,QPGVBCY,US,1,379,227.45,227.45
13,LURLDP PNPJHXMEPN,M,43,PNPJHXMEPN@dell.com,3029418206,QPGVBCY,US,1,9447,83.31,310.76
15,SIQANV QQNKJSURDA,M,66,QQNKJSURDA@dell.com,3354132892,BREQSOA,US,1,3075,33.63,33.63
19,ELUTXG TZIKOOQEMJ,F,59,TZIKOOQEMJ@dell.com,6334227072,RXFPWCJ,US,1,5019,256.30,256.30


---

When we look for customers who ordered multiple times like customer 13, 41 and 44, we see there net ammont increasing cumulatively that is the current and the previouse one.

---

---

### **ROW_NUMBER**

Takes the number of the current row within the partition starting from 1 regardless of the framing

So it doesn't take into acc framing

ROW_NUMBER is a function that we don't have to give or pass a parameter to it, because its going just generate a number for us.
What if we wanna know where our product is positioned in the category by price

Let's do it practically..

---

In [28]:
%%sql

SELECT  prod_id,
        price,
        category,
        ROW_NUMBER() OVER(
            PARTITION BY category
            ORDER BY price
        )
        AS "Position in category by price"
FROM products
ORDER BY "Position in category by price" DESC
LIMIT 50;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store


50 rows affected.


prod_id,price,category,Position in category by price
2216,29.99,9,647
3058,29.99,9,646
4029,29.99,9,645
5118,29.99,9,644
5530,29.99,12,643
6000,29.99,9,643
9974,29.99,4,643
6287,29.99,9,642
3113,29.99,4,642
4055,29.99,12,642


---

### **CONDITIONALS**

What if we wanted only to select something when a certain condition is met?

Recently we have covered the WHERE clause to do filtering.

A condition is a special way of doing the same thing but what is so special about it is we can use it in multiple places in our query.

It's syntax is as follows:

Case statements can be used in multiple places in a query

We can also do multiple things for instance we can change the output of the value like we see in our syntax.

Each return must be a single output

Let's do it practically using our store database.

From the orders table, when customerid = 1, let's return 'my first customer', else return 'not my first customer'

In [31]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
8 rows affected.


table_name,string_agg
categories,"category , categoryname"
cust_hist,"customerid , orderid , prod_id"
customers,"customerid , firstname , lastname , address1 , address2 , city , state , zip , country , region , email , phone , creditcardtype , creditcard , creditcardexpiration , username , password , age , income , gender"
inventory,"prod_id , quan_in_stock , sales"
orderlines,"orderlineid , orderid , prod_id , quantity , orderdate"
orders,"orderid , orderdate , customerid , netamount , tax , totalamount"
products,"prod_id , category , title , actor , price , special , common_prod_id"
reorder,"prod_id , date_low , quan_low , date_reordered , quan_reordered , date_expected"


In [34]:
%%sql

SELECT 
    o.orderid,
    o.customerid,
    CASE
        WHEN o.customerid = 1
        THEN 'my first customer'
        ELSE 'not my first customer'
    END
    AS "customer",
    o.netamount
FROM orders AS o 
ORDER BY o.customerid
LIMIT 50;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
50 rows affected.


orderid,customerid,customer,netamount
10677,2,not my first customer,5.08
2337,3,not my first customer,39.06
9077,6,not my first customer,323.30
6239,7,not my first customer,341.44
1187,11,not my first customer,285.39
3710,12,not my first customer,350.87
379,13,not my first customer,227.45
9447,13,not my first customer,83.31
3075,15,not my first customer,33.63
5019,19,not my first customer,256.30


What if we wanted to filter in a where?

We can use a case statement as if it was inside of a SELECT as we saw above.

It would look something like the following:

In [36]:
%%sql

SELECT 
    o.orderid,
    o.customerid,
    o.netamount
FROM orders AS o
WHERE
    CASE
        WHEN o.customerid > 10
        THEN o.netamount < 100
        ELSE o.netamount > 100
    END 
ORDER BY o.customerid
LIMIT 50;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
50 rows affected.


orderid,customerid,netamount
9077,6,323.30
6239,7,341.44
9447,13,83.31
3075,15,33.63
9748,19,71.29
10281,21,24.32
7075,22,33.86
3486,39,17.24
6017,40,55.13
4919,43,48.33


Above, we are doing a conditional filter 

What we are saying is that if a customer id is beyound 10, then we are only going to show the ones that spent less that 100 dolars, but if they are between 1 and 10, the we are going to show the customers that spent more than 100.

---

We can also use case statements in aggregate functions

Let's do it practically...

In [37]:
%%sql

SELECT 
    SUM(
        CASE
            WHEN o.netamount < 100
            THEN -100
            ELSE o.netamount
        END 
    )
    AS "returns",
    SUM(o.netamount) AS "normal total"
FROM orders AS o
LIMIT 50;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
1 rows affected.


returns,normal total
1906961.25,2371719.74


---

In the above code, we are saying, only sum when the net amount is greater that 100, and then when the net amount is less than 100 we are going to refund them there money

This is business case where somebody comes to you and says, heey, if refunded everyone there money if they spend less than 100 dolars as a gesture of good faith, how would that look like the tatal impact towards our sales.

So we can run a CASE statement that says, when you spend less than 100 dolars, let's just remove 100 from your order, give you 100 dolars back even if you didn't spend 100, otherwise, sum the net amount.

What this would do in the background, any order of less than 100 would be replace by negative 100 and anything else greater than 100, would be added in the sum. Becasue of the negatives, our total we be slightly lower as compared to our normal total.

That is the impact it would have against our normal total.

We can see what are we going to incure in losses if we do this gesture of good faith, how many people spent less than 100 dolars on a singular order if we refunded all oreders that were less than 100 dolars

This can be very bad for business

---

---

### **NULL IF**

What if we wanted to return NULL if a condition is met?

That is wher we use NULLIF

syntax:

When we apply NULLIF to any given statement, we are saying, if val_1 is equal to val_2, return a NULL.

We can use it in a scenario where we are saying if these two values match, i just want NULL, nothing else.

This depends if you are a fun of NULL or not.

If you wanna show nothing at all, NULL is a good option

It works like:

NULLIF(0,0) -- NULL

If we use NULL if 0 and 0, we are gonna get NULL.

NULLIF('ABC', 'DEF') -- ABC

If we use NULLIF ABC, DEF, we are gonna get ABC

The way it works is, if the left hand side matches the right hand side, we are gonna get NULL but if the left hand side does not match the right hand side, then the left hand side gets returned

Hence we can see that we put our column to the left and the value that the column needs to match inorder to return NULL.

We can use NULL in scenarios where we need to fill empty spots with a NULL to avoid devide by ZERO issues and other many scenarios.

---

---

### **VIEWS**

What if we wanted to store the results of a query?

We are running these queries, we are consistently writing new more complex, more full bodied queries that have all these complex concepts in them.

What if we wanted to restore the results of that and reuse them and reuse it at a later time or query against those results?

What if we wanted to query the results of a query?

**Views allow us to store and query previously run queries**

---

#### **Types of VIEWS**

There are two types of views:

* * **MATERIALIZED**
* * **NON-MATERIALIZED** 

**Non-Materialized VIEWS**

In non-materialized view, the query gets re-run each time the view is called on

So we are string a refference to the query but we are re-running it each time that we call the view.

We are not actually storing the results of that, we are storing a refference to the query

**Materialized VIEWS**

Materialized VIEW on the other hand, stores the data physically and periodically updates it when tables change

It we store the data in whatever storage we are using, in our case the drive and periodically updates it when the tables change, so if we INSERT, UPDATE, or DELETE date from our table, periodically the view is going to look at whatever happened and its going to update whatever it stored for us when the table changes.

The whole point of these is, first of all if we have like 2, 3 or 4000 queries, let's say for instance we are working for a big company and we have so many different types of queries, when we have that scenario where we have hundreds and hundreds of queries and we are re-running some of them constantly, or we need to run queries agaist other queries, storing them as views can be a very good option. 

---

---

### **CREATING A VIEW**

We can use the following syntax to create a view:

Using the above syntax, we can Create a VIEW, give it a name and the run the queries that we want to store as VIEWS at the end of the syntax.

What that is going to do, is that, its going to take our query and its going to put it inside what we would call a non-materialized view

**VIEWS are the output of the query we ran**

**VIEWS act like tables, we can query them**

So when we create a view, we can query them.

**VIEWS take very little space to store. We only store the defination of a view not all of the data it returns and this is true for a non-materialized view**

Majority of the time we will want to work with non-materialized views

We may also want to keep materialized views let's say for perfomance reasons, we don't wanna constantly re-ran that query we wanted to update it in the back ground may be because its quite heavy and many more reasons.

So in the majority of the time, we will be working with views which take up very little space.

---

### **UPDATING A VIEW**

Let's say we wanted to update a view by making some small changes to the query

If the view name exist, it will either replace the query if the query we are writing is different, or it will create a completely new one

We can use the following syntax:

---

### **RENAME A VIEW**

What if we wanted to rename a view?

Let's we get into a company and they have 1000 views but they don't have a consistent naming convention hence we have to rename them?

The syntax below, allows us to ALTER a view name and rename it to whatever view name of our interest.

The syntax is as follows:

---

### **DELETING A VIEW**

Let's say we are cleaning a database and there are these views that are not useful at all and we want to delete them, we can us the DROP VIEW function.

We can add a parameter [IF EXISTS] in order to avoid errors if the view doesnot exist in our database.

We can also create multiple drops and add the parameter [IF EXIST], saying, only drop if it exists, if it doesn't exist, don't through an error to us.

If we have many views, let's say 1000, 5000, we can create a scrypt and use the parameter [IF EXSIST] to avoid our scrypt from breaking.

The syntax is as follows:

---

### **USING VIEWS**

Let's say we wanted to get the most recent salary of an employee.

We solved this problem multiple times, first of all using GROUP BY which didn't turn out that well, secondly using WINDOW FUNCTION which went well and we got the right answer only that it had heavy computational needs, and they took time because they ran against each individual record.

Right now we have views, another way to solve the problem.

Let's solve it with views

In [38]:
%sql postgresql://postgres:Kithusi254@localhost:5432/Employees

'Connected: postgres@Employees'

In [40]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [72]:
%%sql

-- Let's say we wanted to get the current salary.

CREATE OR REPLACE VIEW last_salary AS
SELECT  emp.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name)
        AS "name",
        dpt.dept_no,
        dep.dept_name,
        MAX(sa.from_date)
        AS last_salary_change
FROM employees AS emp
INNER JOIN salaries AS sa USING(emp_no)
INNER JOIN dept_emp AS dpt USING(emp_no)
INNER JOIN departments AS dep USING(dept_no)
GROUP BY emp.emp_no, dpt.dept_no, dep.dept_name
ORDER BY emp.emp_no;
-- -------------------------------------------------------------------------------------------------------
--Let's rename the our view
--Let's drop it first
DROP VIEW IF EXISTS last_salary_change;
--Now let's rename it
ALTER VIEW last_salary RENAME TO last_salary_change;
-- --------------------------------------------------------------------------------------------------------
SELECT  lsc.emp_no,
        lsc.name,
        lsc.dept_no,
        lsc.dept_name,
        sa.salary,
        sa.from_date,
        sa.to_date
FROM salaries AS sa
INNER JOIN last_salary_change AS lsc USING(emp_no)
WHERE sa.from_date = lsc.last_salary_change
ORDER BY lsc.emp_no
LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
Done.
Done.
Done.
50 rows affected.


emp_no,name,dept_no,dept_name,salary,from_date,to_date
10001,Georgi Facello,d005,Development,88958,2002-06-22,9999-01-01
10002,Bezalel Simmel,d007,Sales,72527,2001-08-02,9999-01-01
10003,Parto Bamford,d004,Production,43311,2001-12-01,9999-01-01
10004,Chirstian Koblick,d004,Production,74057,2001-11-27,9999-01-01
10005,Kyoichi Maliniak,d003,Human Resources,94692,2001-09-09,9999-01-01
10006,Anneke Preusig,d005,Development,59755,2001-08-02,9999-01-01
10007,Tzvetan Zielinski,d008,Research,88070,2002-02-07,9999-01-01
10008,Saniya Kalloufi,d005,Development,52668,2000-03-10,2000-07-31
10009,Sumant Peac,d006,Quality Management,94409,2002-02-14,9999-01-01
10010,Duangkaew Piveteau,d004,Production,80324,2001-11-23,9999-01-01


---

Another way to do it...

---

In [73]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
11 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
last_salary_change,"emp_no , name , dept_no , dept_name , last_salary_change"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"


In [89]:
%%sql

CREATE OR REPLACE VIEW last_change AS 
SELECT  sa.emp_no,
        MAX(sa.from_date)
        AS "last_salary_change_date"
FROM salaries AS sa
GROUP BY sa.emp_no;
-------------------------------------------------------------------------------------------------------------
SELECT * FROM last_change
LIMIT 20;
DROP VIEW IF EXISTS last_salary_change_;
ALTER VIEW last_change RENAME TO last_salary_change_;
--------------------------------------------------------------------------------------------------------------
SELECT  emp.emp_no,
        CONCAT(emp.first_name, ' ', emp.last_name),
        dpt.dept_no,
        dep.dept_name,
        sa.salary,
        sa.from_date, 
        sa.to_date
FROM last_salary_change_
INNER JOIN employees AS emp USING(emp_no)
INNER JOIN salaries AS sa USING(emp_no)
INNER JOIN dept_emp AS dpt USING(emp_no)
INNER JOIN departments AS dep USING(dept_no)
WHERE sa.from_date = last_salary_change_date
ORDER BY emp_no
LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
Done.
20 rows affected.
Done.
Done.
50 rows affected.


emp_no,concat,dept_no,dept_name,salary,from_date,to_date
10001,Georgi Facello,d005,Development,88958,2002-06-22,9999-01-01
10002,Bezalel Simmel,d007,Sales,72527,2001-08-02,9999-01-01
10003,Parto Bamford,d004,Production,43311,2001-12-01,9999-01-01
10004,Chirstian Koblick,d004,Production,74057,2001-11-27,9999-01-01
10005,Kyoichi Maliniak,d003,Human Resources,94692,2001-09-09,9999-01-01
10006,Anneke Preusig,d005,Development,59755,2001-08-02,9999-01-01
10007,Tzvetan Zielinski,d008,Research,88070,2002-02-07,9999-01-01
10008,Saniya Kalloufi,d005,Development,52668,2000-03-10,2000-07-31
10009,Sumant Peac,d006,Quality Management,94409,2002-02-14,9999-01-01
10010,Duangkaew Piveteau,d004,Production,80324,2001-11-23,9999-01-01


---

In the second method, we are only using our VIEW as a filter to get our desired answer

---

End......